### PR TITLE
[SPARK-38948][TESTS] Fix `DiskRowQueue` leak in `PythonForeachWriterSuite`

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This pr add `try-finally` for  `run` method  of `BufferTester.thread` and call `buffer.close()` in the `finally` block to  ensure the  resources held by `BufferTester.buffer` are released.  

Before this pr, there will be an `DiskRowQueue` resource hold by `BufferTester.buffer` not closed.

### Why are the changes needed?
Minor fix of UT.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Pass GA.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/python/PythonForeachWriterSuite.scala [hunk 1]
```diff
@@ -102,11 +102,15 @@ class PythonForeachWriterSuite extends SparkFunSuite with Eventually with Mockit
...
+        try {
+          while (iterator.hasNext) {
+            outputBuffer.synchronized {
+              outputBuffer += iterator.next().getInt(0)
+            }
+            Thread.sleep(sleepPerRowReadMs)
           }
-          Thread.sleep(sleepPerRowReadMs)
+        } catch {
+          case _: InterruptedException => buffer.close()
```
Comment (LuciferYang) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/python/PythonForeachWriterSuite.scala:113]: `tester.close()` will call `thread.interrupt()`

Comment (HyukjinKwon) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/python/PythonForeachWriterSuite.scala:113]: can we put this in `finally`?

Comment (LuciferYang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/python/PythonForeachWriterSuite.scala]: `finally` is ok

**Review by HyukjinKwon — APPROVED** [review]

**Comment by LuciferYang** [comment]
friendly ping @HyukjinKwon

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**